### DISCUSSION TITLE
Clearable selects

### DISCUSSION THREAD

**Discussion by ArthurDoorgeest** [livewire/flux#346] [Feature Requests] [answered]
The is currently no way to deselect / clear the selected value in a flux:select component.  It would be a usefull addition to add the "clearable" property to flux:select component as well.

**Comment by mnabialek** [comment]
@calebporzio  Any chance to implement this soon?

**Comment by calebporzio** [comment]
Ok, I did an exploration on this.

First, looking at select2 and chosen to see if they offer this and how they handle it in the UI.

Both are not clearable by default but DO offer it.

Here's select2:

![CleanShot 2024-11-01 at 08 03 35](https://github.com/user-attachments/assets/08b1f141-8d25-4ade-831c-495133b0ff81)

Here's chosen:

![CleanShot 2024-11-01 at 08 03 08](https://github.com/user-attachments/assets/324e56f5-5e0f-4759-824e-5884983b382a)

chosen is the better of the two. Select2 has the following flaws IMO:
* The "x" button doesn't hide when nothing is selected
* Clicking "x" triggers the dropdown menu as well
* There is no hover state feedback on the "x"
* The "x" button color is stronger than the chevron down icon "v"

---

That said, I took a crack at an implementation.

The API would be this:

```html
<flux:select variant="listbox" clearable>
    ...
</flux:select>
```

Here's what it looks like:

![CleanShot 2024-11-01 at 08 09 02](https://github.com/user-attachments/assets/8db2e317-e7b6-4b28-8718-b125a25717cc)

My only hesitations with what I have are:
* The "x" icon being a different color than the chevron (to be consistent with the other clearable colors on the input component)
* If I add this, people will expect that you can add "clearable" on the following components:
  * Searchable select (the input in the listbox dropdown)
  * The combobox input
  * The command input

**Comment by calebporzio** [comment, marked as answer]
Ok, I decided to just go for it. Here are some rules for the feature:
- Focusable but not tabbable
- Cursor-pointer
- If clearing an input, focus back to the input after clearing
- Add aria-label to button
- If possible, only show clear button if input or selection is not empty

---

Here it is in action:

https://github.com/user-attachments/assets/0ea97c07-a4c1-4328-8a77-cdbb07c05e36

I'll add documentation for it, and it will be in the next release. Hope this helps. Thanks you for the request.

Oh also, the only input I decided to skip was the combobox because:
* Having two click targets (the expand button and the x-mark button) makes the layout really weird looking
* Sometimes combobox selections aren't even clearable and having the x would be misleading
* People can just select all and press backspace to clear the input

**Comment by Uhasith** [comment]
Can the clearable feature be supported for the ComboBox as well? @calebporzio
